File: scaffold/src/delivery_state_machine.py
"""delivery_state_machine.py — 10 态投递状态机（服务端中枢 · B2-2）

状态严格对齐 applications-list.response.schema.json 的 status 枚举：
  pending_confirm / autofilling / submitted / viewed / contacting /
  interview_invited / interview_done / offer / rejected / closed

转移矩阵对齐 LLD-投递状态机模块 v1.0 §1（ADR-008）：无回退边；
rejected / closed 为终态；所有转移写审计日志。
"""
from __future__ import annotations

import time
from dataclasses import dataclass
from typing import Dict, List, Optional


# 终态：不可再转移
TERMINAL = {"rejected", "closed"}

# 允许转移矩阵（当前态 -> 可到达的下一态集合）
TRANSITIONS: Dict[str, set] = {
    "pending_confirm": {"autofilling"},
    "autofilling": {"submitted", "closed", "pending_confirm"},
    "submitted": {"viewed", "rejected", "closed"},
    "viewed": {"contacting", "rejected", "closed"},
    "contacting": {"interview_invited", "rejected", "closed"},
    "interview_invited": {"interview_done", "rejected", "closed"},
    "interview_done": {"offer", "rejected", "closed"},
    "offer": {"closed"},
    "rejected": set(),
    "closed": set(),
}


class InvalidTransition(Exception):
    """非法状态转移（自环 / 回退 / 终态再转 / 未知态）。"""


@dataclass
class AuditEntry:
    task_id: str
    from_state: str
    to_state: str
    reason: Optional[str]
    ts: int


class DeliveryStateMachine:
    """内存版 10 态投递状态机（演示 + 单测；生产用 MySQL + Redis 幂等键）。"""

    def __init__(self, now=None) -> None:
        self._state: Dict[str, str] = {}
        self._audit: Dict[str, List[AuditEntry]] = {}
        self._now = now or (lambda: int(time.time() * 1000))

    # ---- 生命周期 ----
    def create(self, task_id: str, initial: str = "pending_confirm") -> str:
        if task_id in self._state:
            raise InvalidTransition(f"task {task_id} 已存在")
        if initial != "pending_confirm":
            raise InvalidTransition("初始态必须为 pending_confirm")
        self._state[task_id] = initial
        self._audit[task_id] = []
        return initial

    def state(self, task_id: str) -> str:
        if task_id not in self._state:
            raise KeyError(f"未知 task {task_id}")
        return self._state[task_id]

    def can_transition(self, task_id: str, to_state: str) -> bool:
        cur = self._state.get(task_id)
        if cur is None:
            return False
        return to_state in TRANSITIONS.get(cur, set())

    def transition(self, task_id: str, to_state: str, *, reason: Optional[str] = None) -> str:
        cur = self._state.get(task_id)
        if cur is None:
            raise KeyError(f"未知 task {task_id}")
        if to_state == cur:
            raise InvalidTransition(f"{task_id}: 自环 {cur} -> {to_state} 不允许")
        if to_state not in TRANSITIONS.get(cur, set()):
            raise InvalidTransition(f"{task_id}: 非法转移 {cur} -> {to_state}")
        self._state[task_id] = to_state
        self._audit.setdefault(task_id, []).append(
            AuditEntry(task_id, cur, to_state, reason, self._now()))
        return to_state

    def audit(self, task_id: str) -> List[AuditEntry]:
        return list(self._audit.get(task_id, []))

    @staticmethod
    def idempotency_key(user_id: str, platform: str, job_id: str, apply_date: str) -> str:
        """幂等四元组 (user_id, platform, job_id, apply_date)，对齐 HLD §6.13.2 / ADR-004。"""
        return f"{user_id}|{platform}|{job_id}|{apply_date}"

File: server-java/src/main/java/com/resumeai/module/application/DeliveryStateMachine.java
package com.resumeai.module.application;

import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * 10 态投递状态机（Java 侧中枢 · 对齐 HLD §3.4 / ADR-008 / LLD-投递状态机 v1.0 / Python delivery_state_machine.py）。
 *
 * <p>设计要点（不可偏离）：
 * <ul>
 *   <li>无回退边（如 viewd 不可回 submitted）；</li>
 *   <li>{@code rejected}/{@code closed} 为终态，不可再转移；</li>
 *   <li>所有转移由调用方写审计日志（{@code ApplicationEvent}），本类只做矩阵裁决、不持有数据；</li>
 *   <li>矩阵与 Python 草稿 {@code delivery_state_machine.TRANSITIONS} 严格一致，双语言不漂移。</li>
 * </ul>
 *
 * <p>注意：本类为纯函数式裁决器，状态数据存于 {@code Application} 实体；避免与生产事故相关的
 * 「状态错乱」——任何转移必须先经 {@link #assertTransition(String, String)} 校验。</p>
 */
public final class DeliveryStateMachine {

    /** 终态集合（不可再转移）。 */
    public static final Set<String> TERMINAL = Set.of("rejected", "closed");

    /** 10 态枚举（顺序即 UI 展示/演进序）。 */
    public static final List<String> STATES = List.of(
            "pending_confirm", "autofilling", "submitted", "viewed", "contacting",
            "interview_invited", "interview_done", "offer", "rejected", "closed");

    /** 允许转移矩阵：当前态 -> 可到达的下一态集合（对齐 ADR-008 + LLD §1）。 */
    private static final Map<String, Set<String>> TRANSITIONS = new HashMap<>();

    static {
        TRANSITIONS.put("pending_confirm", Set.of("autofilling"));
        TRANSITIONS.put("autofilling", Set.of("submitted", "closed", "pending_confirm"));
        TRANSITIONS.put("submitted", Set.of("viewed", "rejected", "closed"));
        TRANSITIONS.put("viewed", Set.of("contacting", "rejected", "closed"));
        TRANSITIONS.put("contacting", Set.of("interview_invited", "rejected", "closed"));
        TRANSITIONS.put("interview_invited", Set.of("interview_done", "rejected", "closed"));
        TRANSITIONS.put("interview_done", Set.of("offer", "rejected", "closed"));
        TRANSITIONS.put("offer", Set.of("closed"));
        TRANSITIONS.put("rejected", Set.of());
        TRANSITIONS.put("closed", Set.of());
    }

    private DeliveryStateMachine() {
    }

    /** 是否为终态。 */
    public static boolean isTerminal(String state) {
        return TERMINAL.contains(state);
    }

    /** 当前态是否允许转移到目标态（不含自环判定）。 */
    public static boolean canTransition(String from, String to) {
        if (from == null || to == null) {
            return false;
        }
        return TRANSITIONS.getOrDefault(from, Set.of()).contains(to);
    }

    /**
     * 校验转移合法性，非法抛 {@link IllegalArgumentException}。
     * 覆盖：未知态 / 自环 / 回退边 / 终态再转 / 矩阵外转移。
     */
    public static void assertTransition(String from, String to) {
        if (!STATES.contains(from)) {
            throw new IllegalArgumentException("未知当前态: " + from);
        }
        if (!STATES.contains(to)) {
            throw new IllegalArgumentException("未知目标态: " + to);
        }
        if (from.equals(to)) {
            throw new IllegalArgumentException("自环不允许: " + from + "->" + to);
        }
        if (!canTransition(from, to)) {
            throw new IllegalArgumentException("非法转移(无回退边/矩阵外): " + from + "->" + to);
        }
    }

    /** 业务级幂等四元组（ADR-006 / HLD §6.13.2）：防「同用户同日对同岗」重投命门。 */
    public static String businessIdempotencyKey(String userId, String platformId, String jobId, String applyDate) {
        return userId + "|" + platformId + "|" + jobId + "|" + applyDate;
    }
}
